cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_ctrl_packages.go

```diff
@@ -11,3 +11,119 @@ See the License for the specific language governing permissions and
+	log.Infof("+kapp-controller GetAvailablePackageSummaries")
+
+	// Retrieve the proper parameters from the request
+	namespace := request.GetContext().GetNamespace()
+	cluster := request.GetContext().GetCluster()
+	pageSize := request.GetPaginationOptions().GetPageSize()
+	pageOffset, err := pageOffsetFromPageToken(request.GetPaginationOptions().GetPageToken())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "unable to intepret page token %q: %v", request.GetPaginationOptions().GetPageToken(), err)
+	}
+	// Assume the default cluster if none is specified
+	if cluster == "" {
+		cluster = s.globalPackagingCluster
+	}
+	// fetch all the package metadatas
+	pkgMetadatas, err := s.getPkgMetadatas(ctx, cluster, namespace)
+	if err != nil {
+		return nil, errorByStatus("get", "PackageMetadata", "", err)
+	}
+
+	// paginate the list of results
+	availablePackageSummaries := make([]*corev1.AvailablePackageSummary, len(pkgMetadatas))
+
+	// create the waiting group for processing each item aynchronously
+	var wg sync.WaitGroup
+
+	if len(pkgMetadatas) > 0 {
+		startAt := -1
+		if pageSize > 0 {
+			startAt = int(pageSize) * pageOffset
+		}
+		for i, pkgMetadata := range pkgMetadatas {
+			wg.Add(1)
+			if startAt <= i {
+				go func(i int, pkgMetadata *datapackagingv1alpha1.PackageMetadata) error {
+					defer wg.Done()
+					// fetch the associated packages
+					// Use the field selector to return only Package CRs that match on the spec.refName.
+					fieldSelector := fmt.Sprintf("spec.refName=%s", pkgMetadata.Name)
+					pkgs, err := s.getPkgsWithFieldSelector(ctx, cluster, namespace, fieldSelector)
+					if err != nil {
+						return errorByStatus("get", "Package", pkgMetadata.Name, err)
+					}
+					pkgVersionsMap, err := getPkgVersionsMap(pkgs)
+					if err != nil {
+						return err
+					}
+
+					// generate the availablePackageSummary from the fetched information
+					availablePackageSummary, err := s.getAvailablePackageSummary(pkgMetadata, pkgVersionsMap, cluster)
+					if err != nil {
+						return status.Errorf(codes.Internal, fmt.Sprintf("unable to create the AvailablePackageSummary: %v", err))
+					}
+
+					// append the availablePackageSummary to the slice
+					availablePackageSummaries[i] = availablePackageSummary
+					return nil
+				}(i, pkgMetadata)
+			}
+			// if we've reached the end of the page, stop iterating
+			if pageSize > 0 && len(availablePackageSummaries) == int(pageSize) {
+				break
+			}
+		}
+	}
+	wg.Wait() // Wait until each goroutine has finished
+
+	// TODO(agamez): the slice with make is filled with <nil>, in case of an error in the
+	// i goroutine, the i-th <nil> stub will remain. Check if 'errgroup' works here, but I haven't
+	// been able so far.
+	// An alternative is using channels to perform a fine-grained control... but not sure if it worths
+
+	// filter out <nil> values
```

Comment:
Similar comment to the related PR (for getinstalled? not sure): why are we not returning an error if there was an error while fetching the result, rather than returning incomplete results (which it sounds like the case for <nil> values is).

Reply:
I tend to do it as fail-tolerant as possible; especially, when developing I had some inconsistencies (like in manually deleting CRs) so I opted for being lenient and letting packages have errors.

A reason for that is we currently propagate that error through the chain of calls, so, a maybe temporary inconsistency or error in a single package will prevent the whole packages core API from working until the problem is manually fixed. Is that the intended behavior? I don't know, but, as a user, I don't want to break Kubeapps for a single corrupted package.

Let me leave this TODO as is right now so that we can discuss it in the future. Adding a link to the comment anyway.